### ISSUE TITLE
'power.law.fit' ends up in error "alpha must be greater than one, Invalid value"

### ISSUE THREAD

**Issue by bockthom** [igraph/igraph#1864]
From version 1.2.7 on, the behavior of `igraph::power.law.fit` sometimes results in a weird error:

```R
igraph::power.law.fit(c(8, 8, 3, 0, 0, 0, 0), implementation = "plfit")
Error in power.law.fit.new(x, xmin = xmin, force.continuous = force.continuous) : 
  At other.c:396 : alpha must be greater than one, Invalid value
```
I have had a look at this line in your code already, but the error message printed does not make sense to me, as alpha is not an input parameter for function `power.law.fit` but an output value of this function. Hence, I don't know what to do to fix this. However, I guess this error should be internally handled somewhere in `plfit/plfit.c` but not returned to the user. In previous versions, this function call did not throw an error.

Even if one could think that power law fitting on such a small vector may not be useful at all, it sometimes happens that an input vector derived from a graph exactly looks like this - and the function ends without any error in the previous version of igraph and also ends without error on similar vectors, just a few examples:

`igraph::power.law.fit(c(8, 8, 0, 0, 0, 0), implementation = "plfit")` ends **without error**:
```
$continuous
[1] FALSE

$alpha
[1] Inf

$xmin
[1] 1

$logLik
[1] NaN

$KS.stat
[1] 1.797693e+308

$KS.p
[1] 0
```

`igraph::power.law.fit(c(8, 8, 3, 3, 0, 0), implementation = "plfit")` ends **without error**:
```
$continuous
[1] FALSE

$alpha
[1] Inf

$xmin
[1] 1

$logLik
[1] NaN

$KS.stat
[1] 1.797693e+308

$KS.p
[1] 0
```

but

`igraph::power.law.fit(c(8,8,3, 0, 0, 0, 0), implementation = "plfit")` ends **with error:**:
```
Error in power.law.fit.new(x, xmin = xmin, force.continuous = force.continuous) : 
  At other.c:396 : alpha must be greater than one, Invalid value
```

**Version information**
The above described error only appears in R versions 1.2.7 and later.

**Comment by szhorvat** [comment]
Transferring to the C core repo, as it's reproducible with C/igraph, latest `master` branch.

```

#include <igraph.h>

int main() {
  igraph_vector_t data;
  igraph_plfit_result_t res;

  igraph_vector_init_int_end(&data, -1, 8, 8, 3, 0, 0, 0, 0, -1);
  
  igraph_power_law_fit(&data, &res, -1, false);

  igraph_vector_destroy(&data);

  return 0;
}
```

**Comment by szhorvat** [comment]
This happens inside of `plfit`, namely in `plfit_estimate_alpha_discrete()`, so I'll leave it to you @ntamas.

**Comment by szhorvat** [comment]
I should note that zeros cannot be present in a power law distribution, and make no sense as input. I would have expected `plfit` to remove anything smaller than the automatically determined `xmin` though.

**Comment by bockthom** [comment]
> I should note that zeros cannot be present in a power law distribution, and make no sense as input. I would have expected `plfit` to remove anything smaller than the automatically determined `xmin` though.

Thanks for your comment, this sounds reasonable. The input originates from the `degree` of a graph's vertices, and there can be vertices with a degree of zero. But to check whether the degree distribution follows a power law, it would be ok to add `+1` to each of the input values before passing it to your function, I guess. But on the other hand, one may neglect isolated vertices, so it may be ok to just ignore zeros in the input. But I am not sure what would be more accurate in the actual us-case, this is a question targeting users, I guess. 

Anyway, this is not what this issue mainly is about, as the described error occurs only in some of the cases where zeros are part of the input, and also in previous implementations no error was thrown in such a case.

**Comment by szhorvat** [comment]
@ntamas I believe that the problem is that invalid values are not excluded during the scan to find an estimate for `xmin`, here:

https://github.com/igraph/igraph/blob/master/vendor/plfit/plfit.c#L1162

`xs_copy` is a sorted version of the data. The algorithm starts looking from the first value. Thus, in this case, its first try for `xmin` is `0`, which of course makes no sense.

There are other functions later, such as `plfit_i_logsum_discrete`, which ignore their `xmin` parameter and simply assume that those values have already been skipped.

I am not comfortable making a fix, as the assumptions are not well documented in plfit functions. However, I think this is necessary right after the `px = xs_copy` line in `plfit_discrete()`:

```
    while (*px < 1) {
        px++; m++;
    }
```

This skips over invalid values.

Similar fixes may be needed elsewhere too, e.g. in the continuous case?

However, even after this fix, if there are less than 5 valid input values, we run into problems:

For example, with the input `8, 8, 3, 1` (i.e. only 4 values),

```
/Users/szhorvat/Repos/igraph-main/igraph/vendor/plfit/hzeta.c:462:6: runtime error: index 33 out of bounds for type 'double [33]'
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /Users/szhorvat/Repos/igraph-main/igraph/vendor/plfit/hzeta.c:462:6 in
=================================================================
==76998==ERROR: AddressSanitizer: global-buffer-overflow on address 0x00010f863248 at pc 0x00010e57698c bp 0x7ffee2443690 sp 0x7ffee2443688
READ of size 8 at 0x00010f863248 thread T0
    #0 0x10e57698b in hsl_sf_hZeta0 hzeta.c:462
    #1 0x10e574a93 in hsl_sf_lnhzeta_deriv_tuple_e hzeta.c:584
    #2 0x10e57819c in hsl_sf_lnhzeta_e hzeta.c:636
    #3 0x10e578291 in hsl_sf_lnhzeta hzeta.c:640
    #4 0x10e595149 in plfit_log_likelihood_discrete plfit.c:1079
    #5 0x10e599ac8 in plfit_discrete plfit.c:1200
    #6 0x10f2034c8 in igraph_power_law_fit other.c:378
    #7 0x10d7bbe1e in main test.cpp:10
    #8 0x7fff6e5413d4 in start+0x0 (libdyld.dylib:x86_64+0x163d4)

0x00010f863248 is located 0 bytes to the right of global variable 'hsl_sf_hzeta_eulermaclaurin_series_majorantratios' defined in '/Users/szhorvat/Repos/igraph-main/igraph/vendor/plfit/hzeta.c:138:8' (0x10f863140) of size 264
SUMMARY: AddressSanitizer: global-buffer-overflow hzeta.c:462 in hsl_sf_hZeta0
```

**Comment by ntamas** [comment]
> I have had a look at this line in your code already, but the error message printed does not make sense to me, as alpha is not an input parameter for function power.law.fit but an output value of this function.

Explanation for myself and for posterity: what happens is that `plfit` tries to fit a power-law to the input, gets some nonsensical result where `alpha` is less than 1, and then it tries to calculate the log-likelihood. This is when the error message pops up; basically `plfit` is complaining to its caller (`igraph`) that it makes no sense to calculate the log-likelihood with alpha < 1. I agree that this should not be leaked to the user. Interestingly enough, the upstream `plfit` project does not have this problem (it prints `xmin must be at least 1`, which makes a bit more sense considering that you should not have degree-zero samples in a power-law.

Anyhow, thanks for the bug report, I'll investigate.

**Comment by szhorvat** [comment]
@bockthom :

> Anyway, this is not what this issue mainly is about, as the described error occurs only in some of the cases where zeros are part of the input, and also in previous implementations no error was thrown in such a case.

You need to be careful. No error does not mean that the result was correct. I checked what 1.2.6 does, and it sometimes returns `xmin = 0`, which is invalid.

> But to check whether the degree distribution follows a power law, it would be ok to add `+1` to each of the input values before passing it to your function, I guess.

Be careful: if `x` follows a power law distribution, then `x+1` does not. This is a significant difference in case of a discrete power law.

> But on the other hand, one may neglect isolated vertices, so it may be ok to just ignore zeros in the input.

Ignoring the zeros is a necessity, not an option. A power law distribution with a negative exponent cannot include zero—there is necessarily a cutoff. This is why "xmin" is also a parameter of the distribution. 

Note that virtually all power law distributions you find in the real world will break down when going towards zero (the probability density decays to zero as approaching zero). It is only the tails that are power laws with a good approximation. Thus, to get a good fit, you will always need to throw away the small values from your data. `xmin` estimates the appropriate threshold for this.

In other words, it's of course fine to throw away small values.

----

To be clear, these are not comments on the bug (which will be fixed), just some advice of caution when fitting power laws. Until the issues are resolved, be sure to manually remove any zeros from your data, and to have a reasonable number of data points (as mentioned above, I see problems with fewer than 5, though even for 5, the fitting results would be meaningless).

**Comment by bockthom** [comment]
@szhorvat: 
Thank you very much for your fast and extensive reply. I completely agree with you, everything you mentioned sounds good and reasonable. However, let me briefly respond to two of your comments, for clarifications reasons:

> Be careful: if `x` follows a power law distribution, then `x+1` does not. This is a significant difference in case of a discrete power law.

Agreed. It was not my intention to claim that. However, when figuring out what happens, prior to reporting the bug here, I stumbled over one of the examples in the manual of R igraph, which triggered me to mention the `+1` example:
See the "Examples" section on https://igraph.org/r/html/latest/fit_power_law.html – there the degrees `d` of the vertices of a graph are collected and then a power-law fitting is performed using `d+1`:
```R
d <- degree(g, mode="in")
fit1 <- fit_power_law(d+1, 10)
```
This example is taken from the igraph manual. However, it does not suit your explanations. So, I would like to recommend to update this example in the manual and state correctly how to deal with vertices of degree zero in such a case instead of confusing the reader with this `+1` example.

> to have a reasonable number of data points (as mentioned above, I see problems with fewer than 5, though even for 5, the fitting results would be meaningless).

This comment is absolutely justified. It is absolutely clear that such fitting results are not meaningful. However, it just happens that you perform the power-law fitting on such a small dataset by chance. Just imagine that you analyze the degree distribution of a series of graphs (lets say tens or hundreds of similarly-sized graphs having around hundreds of data points each), where there is one unexpected outlier in this series of graphs which only has 7 vertices of which 4 are isolated – this was how I spotted your bug – in such a case I would rather like to deal with `NA` values or something similar as a result, rather than an error that needs to be handled by me which prevents me from just applying the power-law fitting to all graphs without exception and analyzing the resulting fittings afterwards. Nevertheless, I rely on you that you find a suitable fix for the bug. I hope this helps you to better understand how I spotted the bug and how it happened that there were only roughly 5 data points in my example.

**Comment by ntamas** [comment]
Thank you for the detailed explanation @bockthom; we will find a more suitable example for the manual page (any suggestions @szhorvat ?). As for the bug itself: indeed there were a few changes in the `plfit` implementation between versions 1.2.6 and the current version, and `power.law.fit` now complains loudly about inputs where it used to fail silently before. In this particular case, the problem is caused not only by the zeros in the input vector but also by the fact that the remaining part of the vector does not contain at least three unique integers (which I'd consider a bare minimum necessity to attempt any sort of fit for the remaining part).

I have made a few changes to the underlying `plfit` library today (see [here](https://github.com/ntamas/plfit/commit/181c79041a6171cc4b8b1f2ffad50958b3cb305c)) that are supposed to prevent this from happening. It will take some time until these changes get incorporated into the R interface, though, partly because the R interface is still behind the official C core of igraph and the `plfit` patch was added to the latest version of the C core. We aim to release a new version of the R interface some time early next year that brings it up-to-date with the latest stable version of the C core, so this will probably be the time when this issue gets fixed on the R side.

Until then, the workaround is to remove all values less than 1 from the input vector, and check whether the remaining vector contains at least three distinct integers. If these conditions do not hold, `power.law.fit` will throw an error so it makes no sense calling it.

**Comment by szhorvat** [comment]
> Until then, the workaround is to remove all values less than 1 from the input vector, and check whether the remaining vector contains at least three distinct integers. If these conditions do not hold, `power.law.fit` will throw an error so it makes no sense calling it.

I guess you mean only the condition that there should be three distinct values after removing everything smaller than `xmin >= 1`, right? Having zeros there won't throw an error. By reading your commit, it doesn't, but the above quite reads as if it did.

**Comment by szhorvat** [comment]
@bockthom Thanks for bringing up that example, such comments are useful.

It's late at night, and my mind may not be working correctly, but I think the example is actually fine. Note how the graph is generated:

```
g <- barabasi.game(1000)
```

This is essentially the _directed_ Price model where:
 - some nodes will indeed have zero in-degree
 - all nodes have an extra constant attractiveness of 1 (they get a degree for free)

Here the zeros fit in with the model, and the distribution is actually better approximated by p(k) ~ (1+k)^(-3) than k^(-3).

We can look up the exact distribution on [Wikipedia](https://en.wikipedia.org/wiki/Preferential_attachment#Definition), substitute in the parameters used here, and get the exact formula as p(k) = 4 / ( (k+1)(k+2)(k+3) ).  Put in k=0 and we see that 2/3 of the nodes have zero in-degree, while only 1/6 have an in-degree of 1.

To sum up, the example adds 1 because it is known that in this model it's k+1 that's power-law distributed, not k. The reason is not to get rid of zeros which would interfere with the fitting.

**Comment by szhorvat** [comment]
What I dislike about the typical R documentation format is that is provides little room for explanations. I think we should keep the example as-in but add an explanatory comment.

**Comment by ntamas** [comment]
> What I dislike about the typical R documentation format is that it provides little room for explanations.

I think that the "official" answer for this is that you can put detailed explanations in [vignettes](https://r-pkgs.org/vignettes.html).

**Comment by bockthom** [comment]
> > Until then, the workaround is to remove all values less than 1 from the input vector, and check whether the remaining vector contains at least three distinct integers. If these conditions do not hold, `power.law.fit` will throw an error so it makes no sense calling it.
> 
> I guess you mean only the condition that there should be three distinct values after removing everything smaller than `xmin >= 1`, right? Having zeros there won't throw an error. By reading your commit, it doesn't, but the above quite reads as if it did.

Thanks for providing the workaround, that really helps a lot!
Just one additional question for the future release (whenever it will happen): Will zeros be ignored then or should they be still removed before calling the function then? Will there be an error or not?

**Comment by ntamas** [comment]
Zeros will be skipped and you won't need to remove them. You won't see an error either. Effectively, the function always assumes that there is a power-law distribution "above" `xmin` (inclusive) and that the distribution is unknown or irrelevant below `xmin`, so everything below `xmin`  is silently skipped. As an illustration, this is what I get in the development version after introducing the fix:

```
> power.law.fit(c(8,8,3,3,0,0))
$continuous
[1] FALSE

$alpha
[1] 2.158391

$xmin
[1] 3

$logLik
[1] -8.821607

$KS.stat
[1] 0.312681

$KS.p
[1] 0.8290278
```

**Comment by bockthom** [comment]
@szhorvat sorry for bothering you, but I am still a little bit confused when looking at your example:

> `g <- barabasi.game(1000)`
> This is essentially the _directed_ Price model where:
> 
>     * newly added nodes will indeed have zero in-degree
> 
>     * all nodes have an extra constant attractiveness of 1, so that newly added nodes can attract others too (essentially they get a degree for free—that's the `+ 1`)
> 
> 
> Here the zeros fit in with the model, and the distribution is actually better approximated by p(k) ~ (1+k)^(-3) than by k^(-3). We can look up the exact distribution on [Wikipedia](https://en.wikipedia.org/wiki/Preferential_attachment#Definition), substitute in the parameters used here, and get the exact formula as p(k) = 4 / ( (k+1)(k+2)(k+3) ). Put in k=0 and we see that 2/3 of the nodes have zero in-degree, while only 1/6 have an in-degree of 1.
> 
> To sum up, the example adds 1 because it is known that in this model it's k+1 that's power-law distributed, not k.

According to the [documentation of the `barabasi.game`](https://igraph.org/r/doc/sample_pa.html), this function generates "scale-free graphs according to the Barabasi-Albert model". But when I refer to the Wikipedia page you have mentioned and follow the link to the Barabasi-Albert model, it says [here](https://en.wikipedia.org/wiki/Barab%C3%A1si%E2%80%93Albert_model#Properties) for the degree distribution that  `P (k) ∼ k^(−3)` holds, which is different to what you have stated above (`p(k) ~ (1+k)^(-3)`), if I'm not mistaken.

**Comment by ntamas** [comment]
Wikipedia refers to the _total_ degree of a vertex (i.e. the sum of its in-degree and out-degree), while @szhorvat talks about the out-degree only. Note that in igraph the graph will be directed by default, while Wikipedia treats the graph as undirected. In @szhorvat's interpretation, each vertex has an out-degree of 1 and an in-degree of k, therefore you need to use (1+k) in the formula. In Wikipedia, each vertex has a total degree of k and the directions are not considered.

**Comment by bockthom** [comment]
Ah, I see. Thanks a lot for the clarifications!

**Comment by szhorvat** [comment]
There are many ways to formalize a preferential attachment process. They differ in subtle (or not so subtle) ways. You can read more here: http://networksciencebook.com/chapter/5#degree-dynamics  Pay attention in particular to the grey box describing how the Barabási-Albert model, as originally described, is not fully defined. There are references to papers which define it completely and provide analytic solutions.

In the end, all these model variations produce distributions with a power law tails, but the behaviour at low degrees (d=0, d=1, etc.) may be different.

`barabasi.game` has several parameters and can thus produce many variants of preferential attachment. What is used in the example is known as the Price model and the formula I linked to is the exact solution. https://en.wikipedia.org/wiki/Price%27s_model

**Comment by szhorvat** [comment]
> while Wikipedia treats the graph as undirected

On Wikipedia I was linking to the general page about preferential attachment, which doesn't deal with graphs (directed or undirected), but the general case of balls in urns. I believe the formula for the Yule distribution here: https://en.wikipedia.org/wiki/Preferential_attachment#Definition corresponds to the default directed case of `barabasi.game`. The number of balls is simply the in-degree.

**Comment by ntamas** [comment]
0.9.6 is just around the corner (will be released in a few days) so I'm closing this to clean up the issue tracker.
